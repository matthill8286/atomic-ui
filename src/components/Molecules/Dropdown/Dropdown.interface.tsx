import React from 'react'

export type DropDownListType = 'custom' | 'native'

export interface DropdownProps {
  autoComplete?: string
  initialIndex?: number
  label?: string
  listType?: DropDownListType
  onChange?: (index: number) => void
  options: DropdownOptionProps[]
}

export interface DropdownPropsEnhanced extends DropdownProps {
  showDropdown: boolean
  toggleDropdown: () => void
}

export interface ListProps {
  active: boolean
  isSearchable: boolean
}

export interface StyledDropdownProps {
  hasLabel: boolean
}

export interface StyledDropdownOptionProps {
  active: boolean
  isFocused: boolean
  onClick: React.MouseEventHandler<HTMLLIElement>
}

export interface DropdownOptionProps {
  id?: string
  isFocused?: boolean
  label: string
  url?: string
}

export interface SearchableDropdownProps {
  id?: string
  name?: string
  className?: string
  errorMessage?: string
  initializeOpen?: boolean
  inputValue: string | undefined
  label?: string
  onBlur?: React.FocusEventHandler<HTMLInputElement>
  onInputChange: React.ChangeEventHandler<HTMLInputElement>
  onSelectChange: (item: DropdownOptionProps) => void
  options: DropdownOptionProps[]
  selectedOption: string
  helpText?: string
  inputRef?: React.RefObject<HTMLInputElement>
}
