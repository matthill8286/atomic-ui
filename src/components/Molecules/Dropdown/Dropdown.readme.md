## Dropdown Component ##

Select components are used for collecting user provided information from a list of options.

#### When To Use ####

A dropdown menu for displaying choices - an elegant alternative to the native <select> element.

Utilizing Radio is recommended when there are fewer total options (less than 4 or 5).

#### Changes ####

Placeholder and label had been merged into label prop.
The label behave like a placeholder when the dropdown hasn't a selected value, then after the user select an element it moves into the top left corner.

The prop resetToInitialIndex had been removed, the component will check and set the initialIndex to the currentIndex when it is specified.
