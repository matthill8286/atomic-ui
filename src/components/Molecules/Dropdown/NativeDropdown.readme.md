# NativeDropdown

ariaLabel prop is provided for a11y reasons