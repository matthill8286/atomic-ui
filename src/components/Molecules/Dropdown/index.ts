export * from './Dropdown'
export * from './Dropdown.interface'
export * from './SearchableDropdown'
export * from './NativeDropdown'
