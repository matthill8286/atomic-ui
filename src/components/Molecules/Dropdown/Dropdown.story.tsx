import { action } from '@storybook/addon-actions'
import { number, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'
import { Dropdown } from './Dropdown'
import { DropdownProps } from './Dropdown.interface'
import readme from './Dropdown.readme.md'

const options = [
  { id: '0', label: 'BMW' },
  { id: '1', label: 'Audi' },
  { id: '2', label: 'Mercedes' },
  { id: '3', label: 'Porsche' },
  { id: '4', label: 'Very long text item that is long rly rly long long long' },
]

const commonKnobs = {
  onChange: action('on-change'),
  options,
}

const story = storiesOf('Design System/Molecules/ Dropdown', module)

story.add(
  'Initial',
  () => {
    const knobs: DropdownProps = {
      ...commonKnobs,
      label: text('label', 'type a label'),
    }

    return <Dropdown {...knobs} />
  },
  { info: readme }
)

story.add(
  'Selected',
  () => {
    const knobs: DropdownProps = {
      ...commonKnobs,
      label: text('Label', 'type a label'),
      initialIndex: number('initialIndex', 0),
    }

    return <Dropdown {...knobs} />
  },
  { info: readme }
)

story.add(
  'Native list',
  () => {
    const knobs: DropdownProps = {
      ...commonKnobs,
      label: text('Label', 'type a label'),
      initialIndex: number('initialIndex', 0),
      listType: 'native',
    }

    return <Dropdown {...knobs} />
  },
  { info: readme }
)
