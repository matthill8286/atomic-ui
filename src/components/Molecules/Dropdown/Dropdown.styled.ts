import { Icon } from '@/components/Atoms/Icon'
import { Typo } from '@/components/Atoms/Typography/Typo/Typo'
import { styled } from '@/styles/styled'
import { ListProps, StyledDropdownOptionProps, StyledDropdownProps } from './Dropdown.interface'

export const StyledContainer = styled.div`
  position: relative;
  cursor: pointer;

  & label {
    margin-top: ${({ theme }) => theme.spacing.base.xs};
    line-height: ${({ theme }) => theme.font.lineHeight.xs};
    position: absolute;
    transform-origin: top left;
    transition: color 200ms cubic-bezier(0, 0, 0.2, 1) 0ms, top 200ms cubic-bezier(0, 0, 0.2, 1) 0ms;
  }

  & [data-label-filled='true'] {
    top: 0;
    font-size: 12px;
    transform-origin: top left;
  }

  & [data-label-filled='false'] {
    top: 16px;
    transform-origin: top left;
  }
`

export const StyledDropdown = styled.div<StyledDropdownProps>`
  width: 100%;
  position: relative;
  display: inline-block;
  margin-top: ${({ theme, hasLabel }) => (hasLabel ? theme.font.lineHeight.sm : 0)};
  height: ${({ theme }) => theme.font.lineHeight.sm};
  line-height: ${({ theme }) => theme.font.lineHeight.sm};

  & span {
    position: absolute;
    left: 0;
  }

  & + hr {
    height: 2px;
    width: 100%;
    margin-top: 2px;
  }
`

export const StyledTypo = styled(Typo)`
  max-width: calc(100% - 16px);
`

export const StyledIcon = styled(Icon)`
  pointer-events: none;
  position: absolute;
  top: 0;
  right: 0;
`

export const StyledList = styled.ol<ListProps>`
  ${({ active, isSearchable, theme }) => `
    min-width: 100%;
    max-width: 100%;
    max-height: 320px;
    overflow-y: scroll;
    position: absolute;
    top: ${isSearchable ? `56px` : `32px`};
    z-index: 3;
    list-style: none;
    padding: ${theme.spacing.base.xs} 0;
    margin: 0;
    box-shadow: ${theme.dimension.elevationLevel2};
    background-color: white;
    border-radius: ${theme.dimension.borderRadius1};
    transform: scaleY(0);
    transform: ${active && 'scaleY(1)'};
    transform-origin: 0 0;
    transition: transform ${theme.transition.short} ease;
  `}
`

export const StyledDropdownOption = styled.li<StyledDropdownOptionProps>`
  ${({ isFocused, theme }) => `
    display: block;
    max-width: 100%;
    padding: ${theme.spacing.base.xs} ${theme.spacing.base.sm};
    background: ${theme.color.white};
    hyphens: auto;
    text-align: left;

    &:hover {
      cursor: pointer;
      background: ${theme.color.grey1};
    }

    ${
      isFocused
        ? `
        cursor: pointer;
        background: ${theme.color.grey1};
      `
        : ''
    }
  `}
`

export const StyledSearchableDropdownWrapper = styled.div`
  position: relative;
`

export const StyledFieldWrapper = styled.div<StyledDropdownProps>`
  display: inline-block;
  margin-top: ${({ hasLabel, theme }) => (hasLabel ? theme.spacing.base.md : 0)};
  position: relative;
  width: 100%;
`

export const StyledDropdownSelect = styled.select`
  appearance: none;
  background-color: transparent;
  border: none;
  box-shadow: none;
  cursor: pointer;
  display: block;
  font-size: ${({ theme }) => theme.font.size.sm};
  font-family: ${({ theme }) => theme.font.family.default};
  height: ${({ theme }) => theme.spacing.base.md};
  outline: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 100%;
  z-index: 2;
`
