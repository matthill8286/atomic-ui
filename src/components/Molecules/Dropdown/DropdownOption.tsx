import React from 'react'
import { CopyText } from '@/components/Atoms/Typography'
import { DropdownOptionProps, StyledDropdownOptionProps } from './Dropdown.interface'
import { StyledDropdownOption } from './Dropdown.styled'

export const DropdownOption: React.FC<DropdownOptionProps & StyledDropdownOptionProps> = ({
  active,
  id,
  isFocused = false,
  label,
  onClick,
}) => {
  return (
    <StyledDropdownOption
      active={active}
      aria-selected={isFocused ? 'true' : undefined}
      id={id ?? label}
      isFocused={isFocused}
      onClick={onClick}
      role="option">
      <CopyText tag="strong" color={active ? 'primary' : 'grey5'} fontSize="sm">
        {label}
      </CopyText>
    </StyledDropdownOption>
  )
}

DropdownOption.displayName = 'DropdownOption'
