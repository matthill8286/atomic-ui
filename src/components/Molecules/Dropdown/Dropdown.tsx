import React, { FC, useState } from 'react'
import { InputDivider as Divider } from '@/components/Atoms/Input/InputDivider'
import { CopyText } from '@/components/Atoms/Typography'
import { withDropdownState } from '@/components/Helper/withDropdownState'
import { DropdownPropsEnhanced } from './Dropdown.interface'
import {
  StyledContainer,
  StyledDropdown,
  StyledDropdownSelect,
  StyledFieldWrapper,
  StyledIcon,
  StyledList,
  StyledTypo,
} from './Dropdown.styled'
import { DropdownOption } from './DropdownOption'
import { IconArrow } from '@matthill8286/atomic-icon-library'

const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
  const {
    autoComplete,
    initialIndex,
    label,
    listType = 'custom',
    onChange,
    options = [],
    showDropdown,
    toggleDropdown,
  } = props

  let setupCurrentIndex = -1
  const lastIndexOfOptions = options.length - 1

  if (initialIndex === 0 || !!initialIndex) {
    if (initialIndex >= 0 && initialIndex <= lastIndexOfOptions) {
      setupCurrentIndex = initialIndex
    }
  }

  const [currentIndex, setCurrentIndex] = useState<number>(setupCurrentIndex)
  const [isFilled, setFill] = useState<boolean>(() => currentIndex !== -1)

  const handleItemClick = (index: number) => () => {
    setCurrentIndex(index)
    setFill(true)
    if (onChange) onChange(index)
  }

  const handleSelectChange = (ev: React.ChangeEvent<HTMLSelectElement>): void => {
    const index = options.findIndex(({ id }) => id === ev.target.value)
    if (index !== -1) {
      handleItemClick(index)()
    }
  }

  const selectedItem = options.find((o, i) => i === currentIndex)

  return (
    <StyledContainer onClick={toggleDropdown}>
      {label && (
        <CopyText tag="label" color={isFilled ? 'grey5' : 'grey2'} data-label-filled={isFilled}>
          {label}
        </CopyText>
      )}
      {listType === 'custom' ? (
        <StyledDropdown hasLabel={!!label}>
          {currentIndex !== -1 && (
            <StyledTypo fontSize="sm" limitLines={1} lineHeight="sm" tag="span">
              {options[currentIndex].label}
            </StyledTypo>
          )}
          <StyledIcon width={16} height={16} rotate={showDropdown ? 270 : 90}>
            <IconArrow />
          </StyledIcon>

          <StyledList active={!!showDropdown} isSearchable={false}>
            {options.map((item, index) => (
              <DropdownOption
                key={index}
                isFocused={false}
                active={index === currentIndex}
                label={item.label}
                onClick={handleItemClick(index)}
              />
            ))}
          </StyledList>
        </StyledDropdown>
      ) : (
        <StyledFieldWrapper hasLabel={!!label}>
          <StyledDropdownSelect
            onChange={handleSelectChange}
            {...(selectedItem?.id ? { value: selectedItem.id } : {})}
            {...(autoComplete ? { autoComplete } : {})}>
            {options.map((item, i) => {
              return (
                <option key={item.id || `dropdown-${i}`} value={item.id}>
                  {item.label}
                </option>
              )
            })}
          </StyledDropdownSelect>
          <StyledIcon width={16} height={16} rotate={90}>
            <IconArrow />
          </StyledIcon>
        </StyledFieldWrapper>
      )}
      <Divider color="grey3" />
    </StyledContainer>
  )
}

export const Dropdown = withDropdownState(DropdownWithoutState, { displayName: 'Dropdown' })
