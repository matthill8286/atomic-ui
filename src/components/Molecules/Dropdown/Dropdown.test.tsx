import React from 'react'
import { mountWithTheme, renderWithTheme } from '@/testRenderer'
import { Dropdown } from './Dropdown'
import { StyledDropdownOption, StyledList } from './Dropdown.styled'

describe('Dropdown', () => {
  it('renders', () => {
    const wrapper = renderWithTheme(<Dropdown options={[{ label: 'one' }, { label: 'two' }]} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('renders and show options list', () => {
    const wrapper = mountWithTheme(<Dropdown options={[{ label: 'one' }, { label: 'two' }]} />)
    wrapper.find('DropdownWithoutState').simulate('click')
    const dropDownWithOptions = wrapper.findWhere((item) => item.props().showDropdown === true)
    expect(dropDownWithOptions).toBeDefined()
  })
  it('toggles the dropdown on click', () => {
    const wrapper = mountWithTheme(<Dropdown options={[{ label: 'one' }, { label: 'two' }]} />)
    const button = wrapper.find('div').first()
    button.simulate('click')
    expect(wrapper.find(StyledList).findWhere((item) => item.props().active))
  })
  it('accepts an optional label prop', () => {
    const wrapper = mountWithTheme(
      <Dropdown label="Label text" options={[{ label: 'one' }, { label: 'two' }]} />
    )
    const label = wrapper.find('label')
    expect(label.length).toEqual(1)
    expect(label.text()).toEqual('Label text')
  })
  it('should select the option with given initialIndex prop', () => {
    const wrapper = mountWithTheme(
      <Dropdown
        initialIndex={1}
        label="Label text"
        options={[{ label: 'one' }, { label: 'two' }]}
      />
    )
    expect(
      wrapper.find(StyledDropdownOption).findWhere((item) => {
        const props = item.props()
        return props.active && props.children === 'two'
      })
    )
  })
})
