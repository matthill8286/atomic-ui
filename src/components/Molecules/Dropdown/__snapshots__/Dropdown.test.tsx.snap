// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders 1`] = `
.c7 {
  height: 1px;
  background: #918e8c;
  border-width: 0;
}

.c8 {
  height: 2px;
  width: 100%;
  margin-bottom: 0;
  margin-top: 4px;
}

.c6 {
  color: #272422;
  font-size: 16px;
  line-height: 24px;
  font-family: SourceSansPro,Arial,sans-serif;
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c6 a,
.c6 a:focus,
.c6 a:hover,
.c6 a:active,
.c6 a:visited {
  color: inherit;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 svg {
  cursor: inherit;
  width: 16px;
  height: 16px;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
  fill: currentColor;
}

.c0 {
  position: relative;
  cursor: pointer;
}

.c0 label {
  margin-top: 8px;
  line-height: 20px;
  position: absolute;
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
  -webkit-transition: color 200ms cubic-bezier(0,0,0.2,1) 0ms,top 200ms cubic-bezier(0,0,0.2,1) 0ms;
  transition: color 200ms cubic-bezier(0,0,0.2,1) 0ms,top 200ms cubic-bezier(0,0,0.2,1) 0ms;
}

.c0 [data-label-filled='true'] {
  top: 0;
  font-size: 12px;
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
}

.c0 [data-label-filled='false'] {
  top: 16px;
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
}

.c1 {
  width: 100%;
  position: relative;
  display: inline-block;
  margin-top: 0;
  height: 24px;
  line-height: 24px;
}

.c1 span {
  position: absolute;
  left: 0;
}

.c1 + hr {
  height: 2px;
  width: 100%;
  margin-top: 2px;
}

.c3 {
  pointer-events: none;
  position: absolute;
  top: 0;
  right: 0;
}

.c4 {
  min-width: 100%;
  max-width: 100%;
  max-height: 320px;
  overflow-y: scroll;
  position: absolute;
  top: 32px;
  z-index: 3;
  list-style: none;
  padding: 8px 0;
  margin: 0;
  box-shadow: 0 3px 4px 0 rgba(0,0,0,0.07),0 3px 3px -2px rgba(0,0,0,0.06),0 1px 8px 0 rgba(0,0,0,0.10);
  background-color: white;
  border-radius: 4px;
  -webkit-transform: scaleY(0);
  -ms-transform: scaleY(0);
  transform: scaleY(0);
  -webkit-transform: false;
  -ms-transform: false;
  transform: false;
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transition: -webkit-transform 0.15s ease;
  -webkit-transition: transform 0.15s ease;
  transition: transform 0.15s ease;
}

.c5 {
  display: block;
  max-width: 100%;
  padding: 8px 16px;
  background: #fff;
  -webkit-hyphens: auto;
  -moz-hyphens: auto;
  -ms-hyphens: auto;
  hyphens: auto;
  text-align: left;
}

.c5:hover {
  cursor: pointer;
  background: #f4f4f4;
}

<div>
  <div
    className="c0"
    onClick={[Function]}
  >
    <div
      className="c1"
    >
      <div
        className="c2 c3"
        height={16}
        rotate={90}
        width={16}
      >
        <svg
          height={32}
          viewBox="0 0 24 24"
          width={32}
        >
          <path
            d="M15.78 11.28a.75.75 0 01.22.53v.38a.77.77 0 01-.22.53l-5.14 5.13a.5.5 0 01-.71 0l-.71-.71a.49.49 0 010-.7L13.67 12 9.22 7.56a.5.5 0 010-.71l.71-.7a.5.5 0 01.71 0z"
          />
        </svg>
      </div>
      <ol
        className="c4"
      >
        <li
          className="c5"
          id="one"
          onClick={[Function]}
          role="option"
        >
          <strong
            className="c6"
            color="grey5"
            fontSize="sm"
          >
            one
          </strong>
        </li>
        <li
          className="c5"
          id="two"
          onClick={[Function]}
          role="option"
        >
          <strong
            className="c6"
            color="grey5"
            fontSize="sm"
          >
            two
          </strong>
        </li>
      </ol>
    </div>
    <hr
      className="c7 c8"
      color="grey3"
    />
  </div>
</div>
`;
